Convert generic types early

Emerentius opened the pull request:
Fixes #23.

* Most functions only need a `PathBuf` in the error case. We can just call `as_ref()` early and build a PathBuf in the error case. We can then also leave out the `Into<PathBuf>` bound.
  This is done wherever possible.
* A few APIs always need a `PathBuf`. For those `Into<PathBuf>` is more efficient and the speedup potentially important.
  It does mean that the API differs from `std`. That causes incompatibility for users who write generic functions themselves which follow the `std` convention of taking `AsRef<Path>`. They can't use `fs_err` inside those without code changes, but the necessary changes are very minor.
  This is done for `fs_err::File` constructors and `fs_err::read_dir`.
* Three functions always construct a `PathBuf`, but don't actually need it. They only do because of implementation detaiils.
  All of these have been adapted to require only `AsRef<Path>`. The functions are: `fs_err::{read, read_to_string, write}`

No trait bounds were added, only removed. That makes the change backwards compatible.

matklad commented on src/errors.rs:
```diff
@@ -37,13 +47,17 @@ pub(crate) struct Error {
 
 impl Error {
-    pub fn new<P: Into<PathBuf>>(source: io::Error, kind: ErrorKind, path: P) -> io::Error {
+    pub fn new(source: io::Error, kind: ErrorKind, path: PathBuf) -> io::Error {
+        io::Error::new(source.kind(), Self { kind, source, path })
+    }
+
+    pub fn new_from_ref(source: io::Error, kind: ErrorKind, path: &Path) -> io::Error {
```
Wouldn't something like this work?

```
    pub fn new(source: io::Error, kind: ErrorKind, path: impl Into<PathBuf>) -> io::Error {
        Self::new(source, kind, path.into())
    }

    pub fn new(source: io::Error, kind: ErrorKind, path: PathBuf) -> io::Error {
        io::Error::new(source.kind(), Self { kind, source, path })
    }
```

Emerentius replied:
Sure and I don't think you need the `_new` either. I mentioned that in the OP:
>I made the crate's Error constructors non-generic so that rustc would tell me about all the callsites that needed attention.

There's no need to keep both `new` and `new_from_ref`.

Emerentius replied:
That said I do like the explicitness. It means you couldn't accidentally pass an `&PathBuf` and cause it to compile a new variant needlessly. It really should be either `&Path` or `PathBuf` and nothing else.
`new_from_ref` is more common than `new` and should therefore have a shorter name.

andrewhickman commented:
I think changing the trait bound on `File::new` is acceptable. Pretty much all types that implement `AsRef<Path>` also implement `Into<PathBuf>`, so it should still work for types. In cases where an `impl AsRef<Path>` is passed, then it just needs an `as_ref()` adding to work with this library

Emerentius commented:
I've rebased the PR on top of master, converted the functions which create and immediately consume a `File` to need only `AsRef` and made the error constructors generic again.

It's ready for review. I recommend squashing before merging.

andrewhickman commented:
Thanks!

Emerentius commented:
Off-topic, but I don't wanna open an issue for this.
I've nominated fs-err for [crate of the week](https://users.rust-lang.org/t/crate-of-the-week/2704/821), because I find it has way too few users. I think it has a good chance of appearing in next week's [TWiR](https://this-week-in-rust.org/).
